src/include/detail/linalg/tdb_matrix_with_ids.h

```diff
@@ -0,0 +1,272 @@
+ * @section DESCRIPTION
+ *
+ * Class that provides a matrix interface to two TileDB arrays, one containing
+ * vectors and one IDs.
+ *
+ * @todo Include the right headers for BLAS support.
+ * @todo Refactor ala tdb_partitioned_matrix.h
+ *
+ */
+
+#ifndef TDB_MATRIX_WITH_IDS_H
+#define TDB_MATRIX_WITH_IDS_H
+
+#include "detail/linalg/matrix_with_ids.h"
+#include "detail/linalg/tdb_helpers.h"
+#include "detail/linalg/tdb_matrix.h"
+#include "tdb_defs.h"
+
+/**
+ * Derived from `tdbBlockedMatrix`, which we have derive from `MatrixWithIds`.
+ * Initialized in construction by filling from the TileDB vectors array and the
+ * the TileDB IDs array.
+ */
+template <
+    class T,
+    class IdsType = size_t,
+    class LayoutPolicy = stdx::layout_right,
+    class I = size_t>
+class tdbBlockedMatrixWithIds
+    : public tdbBlockedMatrix<
+          T,
+          LayoutPolicy,
+          I,
+          MatrixWithIds<T, IdsType, LayoutPolicy, I>> {
+  using Base = tdbBlockedMatrix<
+      T,
+      LayoutPolicy,
+      I,
+      MatrixWithIds<T, IdsType, LayoutPolicy, I>>;
+  using Base::Base;
+
+ public:
+  using index_type = typename Base::index_type;
+
+ private:
+  log_timer constructor_timer{"tdbBlockedMatrixWithIds constructor"};
+
+  std::string ids_uri_;
+  std::unique_ptr<tiledb::Array> ids_array_;
+  tiledb::ArraySchema ids_schema_;
+
+ public:
+  tdbBlockedMatrixWithIds(tdbBlockedMatrixWithIds&& rhs) = default;
+
+  /** Default destructor. array will be closed when it goes out of scope */
+  virtual ~tdbBlockedMatrixWithIds() = default;
+
+  /**
+   * @brief Construct a new tdbBlockedMatrixWithIds object, limited to
+   * `upper_bound` vectors. In this case, the `Matrix` is row-major, so the
+   * number of vectors is the number of rows.
+   *
+   * @param ctx The TileDB context to use.
+   * @param uri URI of the TileDB array to read.
+   * @param ids_uri URI of the TileDB array of IDs to read.
+   */
+  tdbBlockedMatrixWithIds(
+      const tiledb::Context& ctx,
+      const std::string& uri,
+      const std::string& ids_uri) noexcept
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_left>)
+      : tdbBlockedMatrixWithIds(ctx, uri, ids_uri, 0, 0, 0, 0, 0, 0) {
+  }
+
+  /**
+   * @brief Construct a new tdbBlockedMatrixWithIds object, limited to
+   * `upper_bound` vectors. In this case, the `Matrix` is column-major, so the
+   * number of vectors is the number of columns.
+   *
+   * @param ctx The TileDB context to use.
+   * @param uri URI of the TileDB array to read.
+   * @param ids_uri URI of the TileDB array of IDs to read.
+   * @param upper_bound The maximum number of vectors to read.
+   * @param temporal_policy The TemporalPolicy to use for reading the array
+   * data.
+   */
+  tdbBlockedMatrixWithIds(
+      const tiledb::Context& ctx,
+      const std::string& uri,
+      const std::string& ids_uri,
+      size_t upper_bound,
+      size_t timestamp = 0)
+    requires(std::is_same_v<LayoutPolicy, stdx::layout_left>)
```

Comment:
I don't exactly know, but [this comment](https://github.com/TileDB-Inc/TileDB-Vector-Search/pull/235#discussion_r1494905625) by @lums658 about `FeatureVectorArray` likely also applies here:

> In general, stdx::layout_right is not well-tested at all. We might want to audit that and make sure it works everywhere as well as stdx::layout_left. The idea at one point was that a feature vector array should be agnostic as to which layout and that we would use a feature vector array everywhere rather than a ColMajorMatrix. I have started putting concepts in many of the function templates as a start to that.
>
> The layout doesn't really matter, because as long as the underlying one-dimensional data has each of the vectors as contiguous data, both row major and column major will have, respectively, rows and columns as contiguous data. We don't ever iterate over a matrix element by element, but rather operator[] pulls out a row or column, i.e., a contiguous vector.
>
> For the same reason, I think the before and after calls in Python to transpose are superfluous.
>
> In case you are wondering why I picked a column major matrix to represent a feature vector array -- in linear algebra a vector is almost universally represented as a column.